### PR TITLE
krt: add new recompute trigger

### PR DESCRIPTION
**Pull request by howardjohn**
Split out of https://github.com/istio/istio/pull/49162

This allows an escape hatch to safely rely on external state from within
krt.

### CODE REVIEW

File: pkg/kube/krt/recomputetrigger.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+// RecomputeTrigger trigger provides an escape hatch to allow krt transformations to depend on external state and recompute
+// correctly when those change.
+// Typically, all state is registered and fetched through krt.Fetch. Through this mechanism, any changes are automatically
+// propagated through the system to dependencies.
+// In some cases, it may not be feasible to get all state into krt; hopefully, this is a temporary state.
+// RecomputeTrigger works around this by allowing an explicit call to recompute a collection; the caller must be sure to call Trigger()
+// any time the state changes.
+type RecomputeTrigger struct {
+	inner StaticSingleton[int]
+	i     int
```
Comment (hzxuzhonghu) [inline pkg/kube/krt/recomputetrigger.go:28]: I canot get to understand the intention of int here even with unit test

Comment (howardjohn) [reply pkg/kube/krt/recomputetrigger.go:28]: I updated with documentation, thanks!

**Review by stevenctl — APPROVED** [review]
lgtm

for ambient, it looks like networks are the only thing that depend on this
in other controllers, it would be tempting to reach for this often

Should probably add some warning signs in the doc here to encourage a strong effort to avoid using it unless totally necessary?

**Comment by howardjohn** [comment]
> lgtm
> 
> for ambient, it looks like networks are the only thing that depend on this in other controllers, it would be tempting to reach for this often
> 
> Should probably add some warning signs in the doc here to encourage a strong effort to avoid using it unless totally necessary?

Definitely agree. We can make network support krt as well, just didn't want to to chew off too much at once

**Pull request merged**